draft-ietf-httpbis-connect-tcp.md

```diff
@@ -62,41 +64,51 @@ This specification describes an alternative mechanism for proxying TCP in HTTP.
 
 # Specification
 
-A template-driven TCP transport proxy for HTTP is identified by a URI Template {{!RFC6570}} containing variables named "target_host" and "target_port".  This URI Template and its variable values MUST meet all the same requirements as for UDP proxying ({{!RFC9298, Section 2}}), and are subject to the same validation rules.  The client MUST substitute the destination host and port number into this template to produce the request URI.
+A template-driven TCP transport proxy for HTTP is identified by a URI Template {{!RFC6570}} containing variables named "target_host" and "target_port".  This URI Template and its variable values MUST meet all the same requirements as for UDP proxying ({{!RFC9298, Section 2}}), and are subject to the same validation rules.  The client MUST substitute the destination host and port number into this template to produce the request URI.  The derived URI serves as the destination of a Capsule Protocol connection using the Upgrade Token "connect-tcp" (see registration in {{new-upgrade-token}}).
+
+When using "connect-tcp", TCP payload data is sent in the payload of new Capsule Types named DATA and FINAL_DATA (see registrations in {{data-capsule}}).  The ordered concatenation of these capsule payloads represents the TCP payload data.  A FINAL_DATA capsule additionally indicates that sender has closed this stream, semantically equivalent to TCP FIN.  After sending a FINAL_DATA capsule, an endpoint MUST NOT send any more DATA or FINAL_DATA capsules on this data stream. (See {{closing-connections}} for related requirements.)
```

Comment:
The strongest argument I see for FINAL_DATA is that it guarantees* correct shutdown handling through semi-compliant HTTP/1.1 intermediaries.  In particular, unclean shutdown signalling is often difficult to implement in HTTP/1.1, where (absent a signal in the Capsule Protocol) it requires injecting a TLS alert (or TCP RST), but TLS and TCP APIs don't always offer this ability.  Previous versions of this draft suggested signalling by sending a truncated capsule, but this requires piercing an abstraction barrier that might reasonably be exposed by a Capsule Protocol implementation.

Even if the client and server are both speaking HTTP/3, there could be an HTTP/1.1 "hop" in between that "bleaches" the stream error.

*Nothing is ever guaranteed.

Reply:
OK, I have moved FINAL_DATA out of this PR and into a separate one (#2949).

I can say that I have personally fought with (1) applications that silently lost data due to truncation and (2) broken Happy Eyeballs with proxies that reported successful connection because the implementation had no way to report failure, so I do believe these issues are real, though maybe not the most common.